### PR TITLE
AFALG: Support AES-GCM

### PR DESCRIPTION
**Pull request by chelsiocrypto**
Extends afalg functionality to another AES variant. AEAD
cipher GCM, will enable TLS to use hardware crypto accelerator
through AF_ALG socket. Support keysize 128,192 ans 256.

Signed-off-by: Atul Gupta <atul.gupta@chelsio.com>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
You will need to sign the CLA for this.  Thanks.

**Comment by chelsiocrypto** [comment]
Pull request has two UN-successful checks, how to get past that
1. cla-check, how to sign the cla?
2. continuous-integration/travis-ci, build failed

**Comment by richsalz** [comment]
You can click on the "details" button for each failing check.   For Travis, you'll then have to click on and view the build logs.  Keep in mind that sometimes the build failures are because of infrastructure hiccups.

**Comment by mattcaswell** [comment]
Travis is complaining because you need to run "make update" and check in any updated files.

**Comment by chelsiocrypto** [comment]
Thanks, make update resolved the Travis error. Is cla missing ok for review for proceed?

**Comment by mattcaswell** [comment]
> Is cla missing ok for review for proceed?

Yes, we can review. But we can't merge until its sorted.

**Comment by dot-asm** [comment]
This would need tests. I also get "Invalid argument" from sendmsg on my computer. It's either bug or it should be guarded by run-time detection. Latter means that if cipher support availability at kernel side is conditional, then engine should detect it upon initialization and *not* register ciphers in question.

File: engines/e_afalg.c [hunk 1]
```diff
@@ -80,13 +80,18 @@ static int afalg_destroy(ENGINE *e);
 static int afalg_init(ENGINE *e);
 static int afalg_finish(ENGINE *e);
 const EVP_CIPHER *afalg_aes_cbc(int nid);
-static cbc_handles *get_cipher_handle(int nid);
+const EVP_CIPHER *afalg_aes_gcm(int nid);
```
Comment (levitte) [inline engines/e_afalg.c]: There's no reason why it should be non-static, please make it static

**Comment by chelsiocrypto** [comment]
I have submitted the CLA, I assume it takes few days for red-check to go for CLA missing. I see Travis build report two failure, this was fine two days back.
Taken care of last review comments.

**Comment by mattcaswell** [comment]
> I have submitted the CLA, I assume it takes few days for red-check to go for CLA missing.

There was a problem with your Corporate CLA (missing schedule A details). Did you not see my email to you on this?

**Comment by chelsiocrypto** [comment]
Oops.. see your mail landed in my Junk-Emails..may be some crazy filter. Will take care and send the copy again.
Thanks
Atul

**Comment by richsalz** [comment]
Closing to trigger the CLA bot.

**Comment by chelsiocrypto** [comment]
last review comment taken care, Any further comments?

**Comment by dot-asm** [comment]
I can't get `util/opensslwrap.sh s_client -engine afalg` to communicate with AES-GCM. Connections gets terminated with "decryption failed or bad record mac". Well, s_server -engine afalg can't communicate either. If TLS is not supposed to work, then engine should be kept away. And whatever that is supposed to work needs test.

**Comment by t8m** [comment]
I do not think we want to add new features to engines as they are deprecated in master.

**Pull request closed without merging**